packages/loader/container-loader/src/containerStateManager.ts

```diff
@@ -0,0 +1,208 @@
+	 * @param specifiedVersion - The specific version of the snapshot to retrieve
+	 * @returns The snapshot requested, or the latest snapshot if no version was specified, plus version ID
+	 */
+	private async fetchSnapshotTree(
+		specifiedVersion: string | undefined,
+	): Promise<{ snapshot?: ISnapshotTree; version?: IVersion | undefined }> {
+		const version = await this.getVersion(specifiedVersion ?? null);
+
+		if (version === undefined && specifiedVersion !== undefined) {
+			// We should have a defined version to load from if specified version requested
+			this.mc.logger.sendErrorEvent({
+				eventName: "NoVersionFoundWhenSpecified",
+				id: specifiedVersion,
+			});
+		}
+		const snapshot = (await this.storageAdapter.getSnapshotTree(version)) ?? undefined;
+
+		if (snapshot === undefined && version !== undefined) {
+			this.mc.logger.sendErrorEvent({ eventName: "getSnapshotTreeFailed", id: version.id });
+		}
+		return { snapshot, version };
+	}
+
+	public async fetchSnapshot(
+		pendingLocalState: IPendingContainerState | undefined,
+		specifiedVersion: string | undefined,
+		supportGetSnapshotApi: boolean | undefined,
+	) {
+		const { snapshot, version } =
+			pendingLocalState === undefined
+				? await this.fetchSnapshotCore(specifiedVersion, supportGetSnapshotApi)
+				: { snapshot: pendingLocalState.baseSnapshot, version: undefined };
+		const snapshotTree: ISnapshotTree | undefined = isInstanceOfISnapshot(snapshot)
+			? snapshot.snapshotTree
+			: snapshot;
+		if (pendingLocalState) {
+			this.snapshot = {
+				tree: pendingLocalState.baseSnapshot,
+				blobs: pendingLocalState.snapshotBlobs,
+			};
+		} else {
+			assert(snapshotTree !== undefined, "Snapshot should exist");
+			// non-interactive clients will not have any pending state we want to save
+			if (this.offlineLoadEnabled && this.isInteractiveClient) {
+				const blobs = await getBlobContentsFromTree(snapshotTree, this.storageAdapter);
+				this.snapshot = { tree: snapshotTree, blobs };
+			}
+		}
+		return { snapshotTree, version };
+	}
+
+	private async fetchSnapshotCore(
+		specifiedVersion: string | undefined,
+		supportGetSnapshotApi: boolean | undefined,
+	): Promise<{ snapshot?: ISnapshot | ISnapshotTree; version?: IVersion }> {
+		if (
+			this.mc.config.getBoolean("Fluid.Container.UseLoadingGroupIdForSnapshotFetch") ===
+				true &&
+			supportGetSnapshotApi === true
+		) {
+			const snapshot = await this.storageAdapter.getSnapshot({
+				versionId: specifiedVersion,
+			});
+			const version: IVersion = {
+				id: snapshot.snapshotTree.id ?? "",
+				treeId: snapshot.snapshotTree.id ?? "",
+			};
+
+			if (snapshot === undefined && specifiedVersion !== undefined) {
+				this.mc.logger.sendErrorEvent({
+					eventName: "getSnapshotTreeFailed",
+					id: version.id,
+				});
+			}
+			return { snapshot, version };
+		}
+		return this.fetchSnapshotTree(specifiedVersion);
+	}
+
+	public setLoadedAttributes(
```

Comment:
this funciton goes away with when you change runtime and resolvedUrl to get functions